### PR TITLE
[MRG] remove redundant ACGT-checking code.

### PR DESCRIPTION
**Pull request by ctb**
This removes `check_and_process_read` from `Hashtable`, and eliminates sequence cleaning from non-bulk-loading code. All such Python-accessible code should use the `cleaned_seq` attribute that is available from both the `ReadParser` code / `Read` objects and the `khmer.utils.broken_paired_reader` code. 

----

See https://github.com/dib-lab/khmer/issues/1541#issuecomment-274828660 for background context.

Prior to this PR, 

* `consume_fasta` & other bulk sequence loading functions automatically upper-cased DNA and ignored any sequence (no matter how long) that contained non-ACGT.
* `trim_on_abundance` and `trim_below_abundance` similarly upper-cased DNA and ignored non-ACGT-containing strings.
* `find_spectral_error_positions` raised an exception on non-ACGT-containing strings.

This PR updates the test code introduced in #1633 to match the new behavior.

This PR also includes #1661.

-----

- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make clean diff-cover` If it introduces new functionality in
  `scripts/` is it tested?
- [x] `make format diff_pylint_report cppcheck doc pydocstyle` Is it well
  formatted?
- [x] Did it change the command-line interface? Only backwards-compatible
  additions are allowed without a major version increment. Changing file
  formats also requires a major version number increment.
- [x] For substantial changes or changes to the command-line interface, is it
  documented in `CHANGELOG.md`? See [keepachangelog](http://keepachangelog.com/)
  for more details.
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?
- [x] Do the changes respect streaming IO? (Are they
  tested for streaming IO?)

### CODE REVIEW

**Comment by ctb** [comment]
Although reading #1491 I'm wondering if we should be using `cleaned_seq` or equivalent in the C++ bulk `consume_fasta` code, instead of `read.sequence`. If so, then we really need to add some tests so that the current code breaks :).

**Comment by ctb** [comment]
Also see https://github.com/dib-lab/khmer/issues/1619#issuecomment-279801266

**Comment by camillescott** [comment]
Also see #1595 -- some more sophisticated alphabet checking was added to the parsers. Basically, you can give it one of the alphabets (as defined in `alphabets.hh`) by name, and it'll use that for its checks; as perhaps implied by that, I'm in favor of moving this sort of checking code into the parsers. Alternatively, the graphs could take a parameter for whether they should run any checks on what they consume (also using the alphabets).

**Comment by camillescott** [comment]
And, looking at #1511 closer, moving that checking code into the parsers is already becoming SOP. Yay.

**Comment by ctb** [comment]
yeppers.

**Comment by ctb** [comment]
@standage note that one thing the partitioning code does is fail to output reads contaiing Ns.

**Comment by standage** [comment]
> note that one thing the partitioning code does is fail to output reads containing Ns

Roger that, good to know.

**Comment by ctb** [comment]
On Mon, Feb 20, 2017 at 08:07:57AM -0800, Daniel Standage wrote:
> > note that one thing the partitioning code does is fail to output reads containing Ns
> 
> Roger that, good to know.

(for now - fixing it in this PR :)

**Comment by ctb** [comment]
I wonder if this code is any faster now that we don't iterate across every sequence 2 or 3 times?

**Comment by betatim** [comment]
Time to review?

**Comment by ctb** [comment]
On Wed, Mar 22, 2017 at 07:47:42AM -0700, Tim Head wrote:
> Time to review?

yes! but this will not be merged until after release.

**Comment by betatim** [comment]
Ran `./scripts/abundance-dist-single.py -x 1e8 -b ecoli_ref-5m.fastq somehisto.hist -k 31 -s` which takes about 120s on my machine both on master and this branch :-/

**Comment by ctb** [comment]
@betatim @luizirber @camillescott @standage this PR is now ready for merge into master and (IMO) should be given high priority for review.  @camillescott's prediction of many merge conflicts was mistaken, thank goodness - only one small conflict to be resolved!

This merge significantly changes the details around ACTGN handling and will require a bump to khmer 3.0.

**Comment by ctb** [comment]
Tests pass!

File: tests/test_sequence_validation.py [hunk 1]
```diff
@@ -152,51 +152,60 @@ def test_read_cleaning_trim_functions_lowercase(Tabletype, reads):
...
-    x = Tabletype(8, PRIMES_1m)
+    x = Countingtype(8, PRIMES_1m)
     for read in reads:
         x.consume(read.cleaned_seq)       # consume cleaned_seq
 
+    # the precise behavior of these functions is all *undefined*
+    # because different hash functions do different things with
+    # non-ACTG characters.  So all we want to do is verify that the
+    # functions execute w/o error on the k-mers before the "bad" DNA,
+    # and don't return positions in the "good" DNA.
```
Comment (standage) [inline tests/test_sequence_validation.py]: The meaning of the phrase *and don't return positions in the "good" DNA* is unclear to me.

**Review by standage — APPROVED** [review]
Looks good to me. As I understand it, khmer now makes no attempt to clean up kmers that are passed to the `sketch.add()` function, and does a `[^ACGT] --> A` conversion for any bulk sequence loading code. Correct?

**Comment by standage** [comment]
Also, does the partitioning code now output reads containing Ns @ctb?

**Comment by ctb** [comment]
@standage yes; re partitioning and ignoring of reads containing N, see: [this removed line](https://github.com/dib-lab/khmer/pull/1590/files#diff-22f0123dd35e0b63ab97e9c8a2c649f9L378)

File: tests/test_sequence_validation.py (hunk 1 above)
Comment (ctb) [inline tests/test_sequence_validation.py]: The behavior of the hash functions on ACTG should all be the same with respect to "good" sequences, and there should be no errors in there to trim at.

**Pull request merged**